Sources/BSON/BSONObjectID.swift

```diff
@@ -1,23 +1,32 @@
+import Foundation
 import NIO
+import NIOConcurrencyHelpers
 
 /// A struct to represent the BSON ObjectID type.
 public struct BSONObjectID: Equatable, Hashable, CustomStringConvertible {
+    internal static let LENGTH = 12
+
     /// This `BSONObjectID`'s data represented as a `String`.
-    public var hex: String { fatalError("ah") }
+    public var hex: String { self.oid.reduce("") { $0 + String(format: "%02x", $1) } }
 
-    public var description: String {
-        self.hex
-    }
+    public var description: String { self.hex }
 
+    /// ObjectID Bytes
     internal let oid: [UInt8]
 
+    internal static let generator = ObjectIDGenerator()
+
     /// Initializes a new `BSONObjectID`.
     public init() {
-        fatalError("ah")
+        self.oid = Self.generator.generate()
     }
 
     /// Initializes a new `BSONObjectID`.
     internal init(_ bytes: [UInt8]) {
+        precondition(
```

Comment:
this precondition will only be triggered in debug mode, which is why we generally use `fatalError`. Is that intentional here?

Reply:
that was a suggestion by me - since this is an internal method and we only call it from places that already validate the length beforehand, it seems like overkill to check twice in release mode. but at the same time it seemed like this method should do *some* kind of validation

Reply:
It is intention in that @kmahar suggested it, I think because this is internal and used currently only when reading out of the ByteBuffer, which has already asserted it has read 12 bytes

Reply:
sgtm